### PR TITLE
feat: tls

### PR DESCRIPTION
**Pull request by DannyS03**
## Context

- Adds content for libp2p TLS
- Part of https://github.com/libp2p/docs/issues/19

## Latest preview

Please view the latest Fleek preview [here](https://bafybeifepdulm6qck4ejlp7sfqdch6bwksx44d5nkzakrg4zjplwdizlw4.on.fleek.co/concepts/secure-comm/tls/).

### CODE REVIEW

**Comment by p-shahi** [comment]
@thomaseizinger can you help review this?

File: content/concepts/secure-comm/tls.md [hunk 1]
```diff
@@ -6,5 +6,129 @@ aliases:
...
+## TLS 1.3 in libp2p
+
+To use TLS 1.3 in libp2p, a peer must first establish a TLS 1.3 connection with another peer
+using the handshake protocol. Once the handshake is complete, the peers can use the encrypted
+connection to exchange data securely and privately.
+
+### Handshake
+
+A TLS 1.3 handshake in libp2p conducts both key sharing and peer authentication
+in a single round trip. Peers must authenticate each other’s identity during the
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:106]: Not sure what key sharing is supposed to mean.
```suggestion
libp2p uses TLS 1.3 handshake to establish a secure connection between two peers. 
Peers authenticate each other’s libp2p peer ID during the handshake.
```

File: content/concepts/secure-comm/tls.md [hunk 2]
```diff
@@ -6,5 +6,129 @@ aliases:
...
+handshake. By extension, servers require client authentication during the TLS
+handshake and will abort if the client fails to authenticate.
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:108]:
```suggestion
```

File: content/concepts/secure-comm/tls.md [hunk 3]
```diff
@@ -6,5 +6,129 @@ aliases:
...
+
+TLS 1.3 is identified during protocol negotiation with the following protocol
+ID: `/tls/1.0.0`.
+
+In libp2p, endpoints authenticate peers by encoding their public key into an X.509
+certificate extension, however, peers can use arbitrary key types and are not constrained
+to those for which the signing of an X.509 certificate is specified.
+
+{{< alert icon="💡" context="note" text="X.509 is an <a class=\"text-muted\" href=\"https://www.itu.int/en/Pages/default.aspx\"> ITU</a> standard defining the format of public key certificates that use asymmetric cryptography for authentication. Certificate extensions were introduced in version 3 of the X.509 standard, which is a field that offers a set of additional attributes that can be included in the certificate to provide more information about the certificate's subject, such as the certificate's intended purpose, the cryptographic algorithms that the certificate uses, and other relevant details."/>}}
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:117]: That's not really relevant, and super old anyway. The RFC is from 2005.

File: content/concepts/secure-comm/tls.md [hunk 4]
```diff
@@ -6,5 +6,129 @@ aliases:
...
+in a single round trip. Peers must authenticate each other’s identity during the
+handshake. By extension, servers require client authentication during the TLS
+handshake and will abort if the client fails to authenticate.
+
+TLS 1.3 is identified during protocol negotiation with the following protocol
+ID: `/tls/1.0.0`.
+
+In libp2p, endpoints authenticate peers by encoding their public key into an X.509
+certificate extension, however, peers can use arbitrary key types and are not constrained
+to those for which the signing of an X.509 certificate is specified.
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:115]:
```suggestion
In libp2p, peer authentication works by encoding the public key into the TLS certificate.
We designed the system such that we can authenticate key types that are usually not
supported by TLS stacks, such as sepc256k1 (which is a key type that can be used for
libp2p keys).
```

File: content/concepts/secure-comm/tls.md [hunk 5]
```diff
@@ -6,5 +6,129 @@ aliases:
...
+For arbitrary key types, an endpoint needs to encode its host key using the
+[libp2p public key extension](https://github.com/libp2p/specs/blob/master/tls/tls.md#libp2p-public-key-extension)
+which is carried in a self-signed certificate to prove ownership of its host key.
+The libp2p public key extension is an X.509 extension with the Object Identier
+`1.3.6.1.4.1.53594.1.1`. A certificate must include the libp2p public key extension
+to be marked valid.
+
+The endpoint generates a signature using its private host key and shares it along with
+its public host key for verification. The signature proves that the peer owned
+the private host key when signing the certificate.
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:128]: That's very specific. I'd be ok with this just living in the specs, and not in the docs.

File: content/concepts/secure-comm/tls.md [hunk 6]
```diff
@@ -6,5 +6,129 @@ aliases:
...
+
+The public host allows the other peer to calculate the peer ID of the endpoint it
+connects to. Failed authentication would immediately terminate the secure connection
+establishment.
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:132]: When processing the certificate, nodes derive the peer ID from the public key that they received,
and clients check that it matches the peer ID of the server they intended to connect to.

**Review by marten-seemann — CHANGES_REQUESTED** [review]

**Review by thomaseizinger — COMMENTED** [review]
Who is the audience of this document?

It starts out very high-level ("What is TLS") but gets detailed fairly quickly, explaining the messages in the handshake. Hence, for someone who's got no idea what TLS is, most of this document is probably overwhelming.

Given that libp2p is just _using_ libp2p in a particular configuration, (i.e. we didn't build it from scratch) I'd recommend the following:

- Retain the super high-level description (i.e. "What is TLS"). This provides some context on what we are talking about.
- From there, link to useful material for people that want to understand TLS in detail (RFC, perhaps Wikipedia for the historical bit)
- Describe that libp2p uses a particular subset, most notably TLS 1.3+
- For any details, we can link to the spec

Hope this helps :)

File: content/concepts/secure-comm/tls.md [hunk 7]
```diff
@@ -6,5 +6,129 @@ aliases:
...
+In TLS 1.3, a client can include all the necessary information, including the key share,
+in the first `ClientHello` message by assuming which key agreement algorithm the
+server will choose due to the limited cipher suites. This saves one round trip as the server
+can generate its key from the first message.
+
+As a result, TLS 1.3 uses a new handshake protocol that allows for faster and more efficient
+establishment of encrypted connections. TLS 1.3 also introduces new features, such as
+certificate compression and support for cryptographic keys larger than 4096 bits.
+
+## TLS 1.3 in libp2p
```
[the remark this answers is not in the document]
Comment (marten-seemann) [reply content/concepts/secure-comm/tls.md:97]: I think calling out the TLS version is high-level enough for a documentation page, but we definitely should link the spec (if we haven't already).

File: content/concepts/secure-comm/tls.md [hunk 8]
```diff
@@ -6,5 +6,51 @@ aliases:
...
+[RFC 8446](https://www.rfc-editor.org/rfc/rfc8446). It brings several
+improvements over TLS 1.2: the latency of a handshake was brought
+down from 2 to 1 network round trips (in the typical case), the privacy
+properties were improved by encrypting the certificates, and the protocol
+was made simplified to reduce implementation complexity.
+
+## TLS 1.3 in libp2p
+
+**libp2p doesn't use TLS versions older than 1.3.** libp2p uses as extended
+version of TLS 1.3, referred to as TLS 1.3+.
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:31]: Not really, we don't touch TLS internals.

File: content/concepts/secure-comm/tls.md [hunk 9]
```diff
@@ -6,5 +6,51 @@ aliases:
...
+
+To use TLS 1.3 in libp2p, a peer must first establish a TLS 1.3 connection with
+another peer using the handshake protocol. Once the handshake is complete, the peers
+can use the encrypted connection to exchange data securely and privately.
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:35]: That's how TLS always works, I find it confusing to call this out explicitly.

File: content/concepts/secure-comm/tls.md [hunk 10]
```diff
@@ -6,5 +6,51 @@ aliases:
...
+
+### Handshake
+
+libp2p uses TLS 1.3 handshake to establish a secure connection between two peers.
+Peers authenticate each other's libp2p peer ID during the handshake.
+
+TLS 1.3 is identified during protocol negotiation with the following protocol
+ID: `/tls/1.0.0`.
+
+In libp2p, peer authentication works by encoding the public key into the TLS certificate.
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:45]:
```suggestion
Peer authentication works by encoding the public key into the TLS certificate.
```

File: content/concepts/secure-comm/tls.md [hunk 11]
```diff
@@ -6,5 +6,51 @@ aliases:
...
+We designed the system to authenticate key types usually not
+supported by TLS stacks, such as sepc256k1 (a key type that can be used for
+libp2p keys).
+
+> X.509 is an [ITU](https://www.itu.int/en/Pages/default.aspx) standard defining the format of public key certificates that use asymmetric cryptography for authentication. Certificate extensions were introduced in version 3 of the X.509 standard, which is a field that offers a set of additional attributes that can be included in the certificate to provide more information about the certificate's subject, such as the certificate's intended purpose, the cryptographic algorithms that the certificate uses, and other relevant details."
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:50]: I'd remove this sentence, it's not really relevant that TLS uses x509 to encode certificates.

File: content/concepts/secure-comm/tls.md [hunk 12]
```diff
@@ -6,5 +6,51 @@ aliases:
...
+
+When processing the TLS certificate, nodes derive the peer ID from the public key that
+they received, and peers check that it matches the peer ID of the server they intend
```
Comment (marten-seemann) [inline content/concepts/secure-comm/tls.md:53]:
```suggestion
they received. The node initiating the connection checks that it matches the peer ID of the node it intended
```

**Review by thomaseizinger — APPROVED** [review]
LGTM! Thanks for incorporating the review :)

File: content/concepts/secure-comm/tls.md [hunk 13]
```diff
@@ -6,5 +6,44 @@ aliases:
...
+properties were improved by encrypting the certificates, and the protocol
+was made simplified to reduce implementation complexity.
+
+## TLS 1.3 in libp2p
+
+**libp2p doesn't use TLS versions older than 1.3.**
+
+### Handshake
+
+libp2p uses TLS 1.3 handshake to establish a secure connection between two peers.
```
Comment (DannyS03) [inline content/concepts/secure-comm/tls.md:34]:
```suggestion
libp2p uses the TLS 1.3 handshake to establish a secure connection between two peers.
```

**Comment by DannyS03** [comment]
thanks @thomaseizinger and @marten-seemann!

available to re-review @marten-seemann

**Review by marten-seemann — COMMENTED** [review]
It would be nice to have an introduction sentence saying that TLS is one of the security handshakes used to secure transports that don't have built-in security (TCP, WebSocket), and also mention that Noise is an alternative option. That would motivate the rest of this document.

Other than that, this looks good to me.

**Comment by DannyS03** [comment]
Added an introductory statement.

**Pull request merged**